File: server-py3/history.py


import json
from datetime import datetime
from pathlib import Path

history_path = Path('history.json')
storage_folder = Path('./uploads')

# ----------------------- msg queue
class MsgList(list):
    def __init__(self):
        super().__init__()
        self.nextid = 0

    def append(self, item):
        super().append(item)
        # print('--hist append:', self.nextid, item)
        self.nextid += 1
        item_id = item.get('data', {}).get('id', 0)
        if self.nextid <= item_id:
            self.nextid = item_id + 1
        # print('  nextid:', self.nextid)

# ----------------------- history
## filter-out expire items @load
def save_history(upload_file_map, message_queue):
    current_time = int(datetime.now().timestamp())
    filtered_files = [
        {
            'name': file['name'],
            'uuid': file['uuid'],
            'size': file['size'],
            'uploadTime': file['uploadTime'],
            'expireTime': file['expireTime'],
        } for file in upload_file_map.values() if file['expireTime'] > current_time
    ]
    filtered_messages = [
        message['data'] for message in message_queue
        if message['data']['type'] != 'file' or message['data']['expire'] > current_time
    ]
    with open(history_path, 'w') as f:
        json.dump({
            'file': filtered_files,
            'receive': filtered_messages
        }, f)


## filter-out expire items @save
def load_history(upload_file_map, message_queue):
    if history_path.exists():
        with open(history_path, 'r') as f:
            history_data = json.load(f)
            current_time = int(datetime.now().timestamp())

            # Load historical files
            for file in history_data['file']:
                if Path(storage_folder / file['uuid']).exists() and file['expireTime'] > current_time:
                    upload_file_map[file['uuid']] = file

            # Load historical messages
            for msg in history_data['receive']:
                if msg['type'] == 'file' and msg['cache'] not in upload_file_map:
                    continue
                message_queue.append({
                    'event': 'receive',
                    'data': {
                        # 'id': len(message_queue),
                        'id': message_queue.nextid,
                        **msg
                    }
                })
                # print('++append', msg)

            # print('---- file map:', upload_file_map)
            # print('--que:', message_queue, id(message_queue))

File: server-go/history.go
package main

/**
*** FILE: history.go
***   handle history.json <===> messageQueue, uploadFileMap
**/

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"time"
)

// file item in File[]
type File struct {
	Name       string `json:"name"`
	UUID       string `json:"uuid"`
	Size       int    `json:"size"`
	UploadTime int64  `json:"uploadTime"`
	ExpireTime int64  `json:"expireTime"`
}

// History represents the entire JSON structure
type History struct {
	File    []File          `json:"file"`
	Receive []ReceiveHolder `json:"receive"` // Interface to handle both text and file types
}

// ----------------- history api
//

var history_path = "history.json"
var storage_folder = "./uploads"

var uploadFileMap = make(map[string]File)
var messageQueue = &PostList{nextid: 0, history_len: config.Server.History}

func _load_history(hist_path string) *History {
	json_str, _ := os.ReadFile(hist_path)

	var payload History
	err := json.Unmarshal([]byte(json_str), &payload)
	if err != nil {
		log.Fatalf("Error parsing JSON: %v", err)
	}

	return &payload
}
func _save_history(hist_path string, payload *History) {
	out, err := json.MarshalIndent(payload, "", "\t")
	if err != nil {
		log.Fatalf("Error marshaling JSON: %v", err)
	}

	if err := os.WriteFile(hist_path, out, 0644); err != nil {
		log.Fatalf("Error writing JSON to file: %v", err)
	}
}

func append_msg(history *History, msg interface{}) {
	var holder ReceiveHolder
	switch v := msg.(type) {
	case TextReceive:
		holder.TextReceive = &v
	case FileReceive:
		holder.FileReceive = &v
	default:
		log.Fatalf("Unknown message type: %T", v)
	}
	history.Receive = append(history.Receive, holder)
}

func pathExists(path string) bool {
	_, err := os.Stat(path)
	return !os.IsNotExist(err)
}

// filter-out file expired or file no-exists msg
func filter_msg(history *History) {
	now := time.Now().Unix()

	// Create a new slice to hold the filtered messages
	var filteredReceive []ReceiveHolder

	for _, receiveItem := range history.Receive {
		if receiveItem.FileReceive != nil {
			fileReceive := receiveItem.FileReceive
			if fileReceive.Expire < now {
				continue // Skip expired files
			}
			if !pathExists(fileReceive.Name) {
				continue // Skip files that do not exist
			}
		}
		filteredReceive = append(filteredReceive, receiveItem)
	}

	// Update the payload with the filtered messages
	history.Receive = filteredReceive
}

// history.json => msg_que,uploadFileMap
func load_history() {
	history := _load_history(history_path)
	current_time := time.Now().Unix()

	// load to uploadFileMap
	for _, file := range history.File {
		if pathExists(filepath.Join(storage_folder, file.UUID)) && file.ExpireTime > current_time {
			uploadFileMap[file.UUID] = file
		} else {
			uuid := file.UUID
			fmt.Println("- del0:", uuid)
			os.Remove(filepath.Join(storage_folder, uuid))
		}
	}

	// load to msg_que
	for _, msg := range history.Receive {
		if msg.Type() == "file" && uploadFileMap[msg.FileReceive.Cache] == (File{}) {
			continue
		}

		// fmt.Println(" --load hist item:", msg)
		messageQueue.Append(&PostEvent{
			Event: "receive",
			Data:  msg,
		})
	}
}

func save_history() {
	current_time := time.Now().Unix()

	filteredFiles := []File{}
	for _, file := range uploadFileMap {
		if file.ExpireTime > current_time {
			filteredFiles = append(filteredFiles, file)
		}
	}

	filteredMessages := []ReceiveHolder{}
	for _, message := range messageQueue.List {
		if message.Data.Type() != "file" || message.Data.FileReceive.Expire > current_time {
			filteredMessages = append(filteredMessages, message.Data)
		}
	}
	historyData := History{
		File:    filteredFiles,
		Receive: filteredMessages,
	}

	file, _ := json.MarshalIndent(historyData, "", "\t")
	_ = os.WriteFile(history_path, file, 0644)
}

func get_all_msg(history *History) []interface{} {
	var msg_list []interface{}

	for _, receiveItem := range history.Receive {
		if receiveItem.TextReceive != nil {
			msg_list = append(msg_list, *receiveItem.TextReceive)
		} else if receiveItem.FileReceive != nil {
			msg_list = append(msg_list, *receiveItem.FileReceive)
		}
	}

	return msg_list
}

/*

func main() {
	// JSON data
	jsonData, err := os.ReadFile("history.test.json")
	if err != nil {
		log.Fatalf("Error reading JSON file: %v", err)
	}

	// Parse the JSON
	var payload Payload
	err := json.Unmarshal([]byte(jsonData), &payload)
	if err != nil {
		log.Fatalf("Error parsing JSON: %v", err)
	}

	fmt.Println("--litter.dump:")
	litter.Dump(payload)

	// Store file data in file_list
	file_list := payload.File

	// Store receive data as concrete types in msg_list
	var msg_list []interface{} // Using interface{} to hold both TextReceive and FileReceive

	for _, receiveItem := range payload.Receive {
		if receiveItem.TextReceive != nil {
			msg_list = append(msg_list, *receiveItem.TextReceive)
		} else if receiveItem.FileReceive != nil {
			msg_list = append(msg_list, *receiveItem.FileReceive)
		}
	}

	// Print file_list and msg_list
	fmt.Println("File List:")
	for _, file := range file_list {
		fmt.Printf("%+v\n", file)
	}

	fmt.Println("\nMessage List:")
	for _, msg := range msg_list {
		switch v := msg.(type) {
		case TextReceive:
			fmt.Printf("Text Message: %+v\n", v)
		case FileReceive:
			fmt.Printf("File Message: %+v\n", v)
		}
	}

	// out, err := json.Marshal(payload)
	out, err := json.MarshalIndent(payload, "", "\t")
	fmt.Println("---json:", string(out))

	msg_list = append(msg_list, TextReceive{ReceiveBase: ReceiveBase{ID: 123, Type: "text", Room: "room1"}, Content: "ping"})
	out, err = json.MarshalIndent(msg_list, "", "\t")
	fmt.Println("---json:", string(out))
}


//*/
